src/data/fin.lean

```diff
@@ -308,6 +386,16 @@ lemma cast_le_injective {n₁ n₂ : ℕ} (h : n₁ ≤ n₂) : injective (fin.c
 lemma cast_succ_injective (n : ℕ) : injective (@fin.cast_succ n) :=
 cast_le_injective (le_add_right n 1)
 
+@[simp] lemma succ_above_below (p : fin (n + 1)) (i : fin n) (h : i.val < p.val) :
```

Comment:
I'm skeptical that this and `succ_above_above` should really be `simp` lemmas since I don't expect `simp` to be able to synthesize the proof `h` often.

Reply:
Should `succ_above` be reformulated as:
```
def succ_above (p : fin (n+1)) (i : fin n) : fin (n+1) :=
if i.cast_succ < p then i.cast_succ else i.succ
```
That way we avoid working on the val field (or subtype value, once the other PR is done), and can provide proofs of the inequality in more ways.